File: services/sparkd/src/sparkd/signature.py
"""La signature d'un geste : forme canonique et vérification.

@spec docs/BACKLOG.md#SPK-40 · docs/DAT.md §36.3 (où la signature est produite
      décide de ce qu'elle vaut), §36.4 (deux classes de lignes),
      §36.10 (le contrat), §36.10.2 (SSHSIG, mesuré), §36.10.3 (ce qui est
      signé), §36.10.4 (où la vérification a lieu), §36.10.5 (les clés
      autorisées) · §36.5 (les pièges de sérialisation)

**Ce que ce module N'EST PAS.** Ce n'est pas de l'authentification : la clé volée
signe, et l'arbitrage de SPK-35 (§45.4) l'a établi. Ce qu'une signature prouve,
c'est qu'un geste inscrit au journal a bien été DEMANDÉ, et n'a pas été fabriqué
par la Forge après coup.

Conséquence directe, et elle commande le contrat : **une requête non signée reste
acceptée**. Refuser faute de signature ferait de ce mécanisme un contrôle
d'accès. En revanche une signature PRÉSENTE et invalide est refusée — l'inscrire
ferait mentir le journal, qui affirmerait une preuve qu'il n'a pas.

**La vérification faite ici n'est pas la vraie.** Elle est faite par la machine
qu'on soupçonne : elle attrape l'erreur et le bruit, pas l'adversaire qui a root.
Qui audite rejoue la vérification AILLEURS, avec les octets et la signature que
le journal conserve — même principe que l'ancre du §36.2.
"""

from __future__ import annotations

import base64
import json
import subprocess
import tempfile
from pathlib import Path

#: Version de la forme sérialisée du §36.10.3. Toute évolution est une RUPTURE et
#: se traite par une version nouvelle, jamais par un changement en place (§36.9.2).
VERSION = "sshsig-v1"

#: Espace de noms SSHSIG. Il n'est PAS décoratif : sans lui, une signature
#: produite par le responsable pour un autre usage — un commit, un courriel —
#: serait rejouable ici. MESURÉ : un espace de noms différent rend 255.
NAMESPACE = "spark-audit"

#: Champs retenus, et eux seuls (§36.10.3).
CHAMPS = ("action", "actor", "body", "method", "path", "ts")


class SignatureError(Exception):
    """Ce qui empêche de vérifier, dit en clair et sans le confondre avec un
    refus d'accès."""

    def __init__(self, message: str, motif: str):
        super().__init__(message)
        #: Le motif est stable ; le message est pour l'humain.
        self.motif = motif


def canonique(intention: dict) -> bytes:
    """Les octets du §36.10.3, figés.

    JSON, séparateurs sans espace, clés triées par ordre d'octets,
    `ensure_ascii`, UTF-8. Une valeur absente est sérialisée `null`, JAMAIS
    omise : omettre une clé produirait deux octets différents pour deux
    intentions équivalentes (§36.9.2).
    """
    retenu = {champ: intention.get(champ) for champ in CHAMPS}
    return json.dumps(retenu, sort_keys=True, ensure_ascii=True,
                      separators=(",", ":")).encode("utf-8")


def decrit_bien(octets: bytes, *, method: str, path: str, actor: str) -> bool:
    """Les octets signés décrivent-ils la requête REÇUE ?

    Sans ce contrôle, on signerait n'importe quoi et on l'attacherait à n'importe
    quel geste : la signature serait valide, et attachée à autre chose.
    """
    try:
        vu = json.loads(octets.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return False
    return (vu.get("method") == method and vu.get("path") == path
            and vu.get("actor") == actor)


def _armure(signature: str) -> str:
    """Rend l'armure PEM si elle manque.

    L'en-tête voyage sur une ligne (§36.10.7) ; `ssh-keygen` exige l'armure.
    """
    propre = signature.strip()
    if propre.startswith("-----BEGIN SSH SIGNATURE-----"):
        return propre if propre.endswith("\n") else propre + "\n"
    corps = "".join(propre.split())
    lignes = [corps[i:i + 70] for i in range(0, len(corps), 70)]
    return ("-----BEGIN SSH SIGNATURE-----\n" + "\n".join(lignes)
            + "\n-----END SSH SIGNATURE-----\n")


def signataires(chemin: str | Path | None) -> Path | None:
    """Le fichier `allowed_signers`, s'il existe ET porte quelque chose.

    Absent ou vide, la fonction se DÉSACTIVE — elle ne tombe pas en panne
    (§36.10.5). Une signature qu'on ne peut rattacher à personne ne prouve rien.
    """
    if not chemin:
        return None
    fichier = Path(chemin)
    if not fichier.is_file() or not fichier.read_text("utf-8").strip():
        return None
    return fichier


def verifier(octets: bytes, signature: str, identite: str,
             allowed_signers: str | Path | None,
             *, executer=subprocess.run) -> None:
    """Vérifie une signature SSHSIG. Lève `SignatureError` si elle ne tient pas.

    MESURÉ sur OpenSSH 8.9p1 (§36.10.2) : `-Y verify` rend 0 quand tout tient, et
    255 sur un message altéré, une identité inconnue, un espace de noms différent
    ou une clé absente de la liste. Le code de sortie fait donc foi, et le
    message d'erreur d'OpenSSH est repris tel quel plutôt que reformulé — il
    nomme précisément lequel des quatre cas s'est produit.
    """
    fichier = signataires(allowed_signers)
    if fichier is None:
        raise SignatureError(
            "Aucun fichier de signataires autorisés n'est configuré sur cette "
            "Forge : une signature ne peut être rattachée à personne, et "
            "l'inscrire affirmerait une preuve qu'elle n'a pas. Réglez "
            "SPARKD_ALLOWED_SIGNERS, ou n'envoyez pas de signature.",
            "signataires_absents")

    with tempfile.TemporaryDirectory() as dossier:
        chemin_sig = Path(dossier) / "geste.sig"
        chemin_sig.write_text(_armure(signature), "utf-8")
        vu = executer(
            ["ssh-keygen", "-Y", "verify", "-f", str(fichier),
             "-I", identite, "-n", NAMESPACE, "-s", str(chemin_sig)],
            input=octets, capture_output=True, timeout=20)

    if vu.returncode != 0:
        detail = (vu.stderr or b"").decode("utf-8", "replace").strip()
        raise SignatureError(
            f"La signature de ce geste ne se vérifie pas : {detail or 'refusée'}. "
            "Elle n'est pas inscrite — un journal qui porte une signature "
            "invalide affirme une preuve qu'il n'a pas.",
            "signature_invalide")


def entete_present(entetes) -> bool:
    """Une signature accompagne-t-elle cette requête ?"""
    return bool(entetes.get("x-spark-signature"))


def lire_entetes(entetes) -> tuple[str, bytes]:
    """Extrait la signature et les octets signés d'une requête.

    Les octets VOYAGENT plutôt que d'être reconstruits (§36.10.7) : reconstruire
    supposerait que deux implémentations sérialisent à l'octet près, pour
    toujours — le premier piège du §36.5, doublé.
    """
    signature = (entetes.get("x-spark-signature") or "").strip()
    brut = (entetes.get("x-spark-signed") or "").strip()
    if not brut:
        raise SignatureError(
            "Une signature accompagne ce geste, mais pas les octets qu'elle "
            "couvre : il n'y a rien à vérifier. Envoyez X-Spark-Signed.",
            "octets_absents")
    try:
        octets = base64.b64decode(brut, validate=True)
    except Exception:
        raise SignatureError(
            "Les octets signés ne sont pas du base64 valide.",
            "octets_illisibles") from None
    return signature, octets

File: apps/webui/host/signature.js
/**
 * La console signe l'intention d'un geste, par l'agent du responsable.
 *
 * @spec docs/BACKLOG.md#SPK-40 · docs/DAT.md §36.3 (où la signature est produite
 *       décide de ce qu'elle vaut), §36.10.1 (ce que l'unité n'est PAS),
 *       §36.10.3 (la forme canonique), §36.10.7 (les en-têtes),
 *       §36.10.8 (côté console) · §21.6.2 (l'identité déclarée) ·
 *       docs/DESIGN_SYSTEM.md §14.7 (pas de jeton technique à l'écran)
 *
 * **Le point qui décide de ce module.** La commande désigne la clé **PUBLIQUE**,
 * jamais la privée. MESURÉ le 2026-08-21 sur OpenSSH 8.9p1 : avec un agent qui
 * détient la clé, `ssh-keygen -Y sign -f cle.pub` signe **sans que la clé privée
 * soit sur le disque** — retirée du disque, la signature est produite et se
 * vérifie. C'est la propriété du §36.3, et elle vaut ici pour la console
 * elle-même, qui signe sans jamais tenir le secret.
 *
 * **Ne pas pouvoir signer n'empêche JAMAIS le geste** (§36.10.8). Refuser d'agir
 * faute de signature ferait de ce mécanisme un contrôle d'accès, et un exploitant
 * dont l'agent vient de se vider ne doit pas découvrir que son produit s'est
 * verrouillé. L'échec est dit, jamais tu.
 */

import { spawn } from 'node:child_process';
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

/** La même version que la Forge attend (`sparkd.signature.VERSION`). */
export const VERSION = 'sshsig-v1';

/** Le même espace de noms, et il n'est pas décoratif (§36.10.2). */
export const NAMESPACE = 'spark-audit';

/** Champs retenus, et eux seuls — l'ordre vient du tri, pas d'ici (§36.10.3). */
export const CHAMPS = ['action', 'actor', 'body', 'method', 'path', 'ts'];

/** Motifs d'un échec de signature. Ils sont STABLES, et ils ne sont QUE des
 *  jetons : c'est sous cette forme qu'ils voyagent jusqu'au navigateur, un
 *  en-tête HTTP ne transportant pas d'accent (§36.10.9).
 *
 *  **La phrase française ne vit pas ici** mais dans le vocabulaire de la
 *  console — `apps/webui/src/components/tokens.js` —, à un seul endroit
 *  (`docs/DESIGN_SYSTEM.md` §12.5). Une seconde copie ici finirait par diverger,
 *  et un motif sans phrase serait un échec TU : exactement ce que le §36.10.8
 *  interdit. Le §14.7 interdit de son côté que le jeton lui-même atteigne
 *  l'écran. */
export const SANS_CLE = 'sans_cle';
export const AGENT_MUET = 'agent_muet';
export const ECHEC = 'echec_signature';

/** L'ensemble des motifs, figé. La console garde que sa table les couvre TOUS. */
export const MOTIFS = Object.freeze([SANS_CLE, AGENT_MUET, ECHEC]);

/**
 * Les octets du §36.10.3, figés.
 *
 * `JSON.stringify` avec des clés TRIÉES, sans espace. Une valeur absente est
 * sérialisée `null`, jamais omise : omettre une clé produirait deux octets
 * différents pour deux intentions équivalentes, et la Forge refuserait une
 * signature parfaitement valide.
 */
export function canonique(intention) {
  const retenu = {};
  for (const champ of [...CHAMPS].sort()) {
    const valeur = intention?.[champ];
    retenu[champ] = valeur === undefined ? null : valeur;
  }
  return Buffer.from(JSON.stringify(retenu), 'utf8');
}

/**
 * Traduit l'échec d'`ssh-keygen` en ce qui MANQUE réellement.
 *
 * MESURÉ : sans agent et sans fichier privé, OpenSSH rend « Load key "…" : No
 * such file or directory ». Ce message nomme un fichier que l'exploitant n'a pas
 * demandé, et le §14.7 interdit un jeton technique à l'écran.
 */
export function classer(code, erreurs = '') {
  if (code === 0) return null;
  return /No such file|Could not open|not found/i.test(String(erreurs))
    ? AGENT_MUET : ECHEC;
}

function lancer(programme, args, spawnFn) {
  return new Promise((resoudre) => {
    const enfant = spawnFn(programme, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    let sortie = '';
    let erreurs = '';
    enfant.stdout?.on('data', (b) => { sortie += b.toString('utf8'); });
    enfant.stderr?.on('data', (b) => { erreurs += b.toString('utf8'); });
    // `close` et non `exit` : `exit` précède le drainage de stdout, et le relevé
    // perdrait des octets en silence (§37.6 ter, mesuré).
    enfant.on('close', (code) => resoudre({ code: code ?? 0, sortie, erreurs }));
    enfant.on('error', () => resoudre({ code: 255, sortie: '', erreurs: 'introuvable' }));
  });
}

/**
 * Signe une intention. Rend `{ signature, signed }` ou `{ motif }`.
 *
 * Ne LÈVE jamais : un geste qui ne peut pas être signé part quand même
 * (§36.10.8). Le motif remonte pour être DIT, pas pour arrêter l'appelant.
 */
export async function signer(intention, { signingKey = null, doublon = null,
                                          spawn: spawnFn = spawn } = {}) {
  const octets = canonique(intention);
  const signed = octets.toString('base64');

  if (!doublon && !signingKey) return { motif: SANS_CLE, signed };

  const dossier = await mkdtemp(join(tmpdir(), 'spark-sign-'));
  try {
    const fichier = join(dossier, 'intention');
    await writeFile(fichier, octets);

    // §36.10.8 : le doublon remplace la COMMANDE, pas le mécanisme — même motif
    // qu'au §37.4.2 bis. Tout le reste du chemin est celui de la production.
    const [programme, ...args] = doublon
      ? ['sh', '-c', doublon, fichier]
      : ['ssh-keygen', '-Y', 'sign', '-f', signingKey, '-n', NAMESPACE, fichier];

    const vu = await lancer(programme, args, spawnFn);
    const motif = classer(vu.code, vu.erreurs);
    if (motif) return { motif, signed };

    const armure = await readFile(`${fichier}.sig`, 'utf8').catch(() => vu.sortie);
    if (!armure.trim()) return { motif: ECHEC, signed };

    // L'en-tête porte la signature sur UNE ligne (§36.10.7) : l'armure est
    // retirée ici, et la Forge la remet. Un en-tête HTTP ne transporte pas de
    // saut de ligne.
    const surUneLigne = armure.split('\n')
      .filter((l) => l && !l.startsWith('-----')).join('');
    return { signature: surUneLigne, signed };
  } finally {
    await rm(dossier, { recursive: true, force: true });
  }
}

/**
 * Les deux en-têtes du §36.10.7, prêts à poser sur une requête.
 *
 * Rend un objet VIDE quand rien n'a pu être signé : envoyer une signature vide
 * ferait refuser le geste par la Forge en `422`, alors que le §36.10.1 veut
 * précisément qu'il passe.
 */
export function entetes(resultat) {
  if (!resultat?.signature) return {};
  return {
    'x-spark-signature': resultat.signature,
    'x-spark-signed': resultat.signed,
  };
}
